### DISCUSSION TITLE
Private model attributes

### DISCUSSION THREAD

**Discussion by sempervalensdev** [strapi/strapi#7835] [Legacy Topics]
I'd like to start with thanks to the Strapi team, great product!

I'm attempting to set model attributes to private within my API definitions. My understanding is that there is no way to do this through the Strapi Admin backend, so I am editing the `/api/{modelName]/models/{modelName}.settings.json` for each model where I wish to exclude certain attributes, adding `"private": true` for each attribute that I wish to hide. This is working as expected, the associated attribute(s) aren't returned when I query the endpoint - yay!

However, when I make a change to the associated model using the Strapi Admin backend, my change is removed. I have a large number of models that I'm working on, making minor tweaks as I build out my application. It's frustrating to continually have to go in and relabel all of the private attributes as private. 

So, am I missing something about the correct way to set a model attribute as private?

**Comment by derrickmehaffy** [comment]
You are not and it's an issue with the Content-type builder, can you open up a bug report instead of a discussion so this can be properly tracked?

**Reply by sempervalensdev** [reply]
Not a problem, thanks for the quick response.

**Reply by sempervalensdev** [reply]
In the course of documenting the bug, I created a new Strapi instance to capture code snippets of simple content-type models, and discovered that v3.1.4 has the functionality in the field editor to mark fields a private and properly retain the configuration. My application development instance is also at v3.1.4, but I arrived there through following the migration guide steps to upgrade from v3.0.5 about a week ago. So it appears that the behavior that I'm seeing is due to a faulty migration. I'll revert back to my v3.0.5 instance and see if I can figure out what went south in my migration that caused me to not get the correct build of the Strapi Admin app.

**Reply by sempervalensdev** [reply]
After digging into this further, I found that the migration doesn't seem to be a problem because I can mark direct fields as private in the Content-Types Builder. The problem that I am seeing is because I am attempting to set a **relation** as private, rather than a direct field. It appears that this isn't possible in the Content-Types Builder. Setting the relation as private by adding the `"private": true` to the attribute in the underlying model works as expected. But subsequent changes to the model will result in the `"private": true` being removed. I'm guessing that this is by design. Can you confirm?

**Reply by derrickmehaffy** [reply]
It is for now as we don't technically support private relations at the moment (nor do we support them being set as required)

It's something that -might- work but we have no idea if it will break stuff. (As you have found out)

**Reply by sempervalensdev** [reply]
Thanks, I appreciate the confirmation. I'll change course and select one of the direct field types for my use case. I'd rather refactor my 34 content-types to use a supported private field type than have to remember to set the property for the attribute back to private every time I or a team member modifies the model.
